docs/exts/provider_init_hack.py

```diff
@@ -0,0 +1,58 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+"""
+Bugs in sphinx-autoapi using metaclasses prevent us from upgrading to 1.3
+which has implicit namespace support. Until that time, we make it look
+like a real package for building docs
+"""
+import os
+import shutil
+
+from sphinx.application import Sphinx
+
+ROOT_PROJECT_DIR = os.path.abspath(
+    os.path.join(os.path.dirname(os.path.realpath(__file__)), os.pardir, os.pardir)
+)
+
+PROVIDER_INIT_FILE = os.path.join(ROOT_PROJECT_DIR, "airflow", "providers", "__init__.py")
+
+
+def _create_init_py(app, config):
+    del app
+    del config
+    with open(PROVIDER_INIT_FILE, "wt"):
+        pass
+
+
+def _delete_init_py(app, exception):
+    del app
+    del exception
+    shutil.rmtree(PROVIDER_INIT_FILE, ignore_errors=True)
```

Comment:
Rmtree does not remove the file (it only removes directories). There was a bug in the original code which I fixed here https://github.com/apache/airflow/pull/12082/files#diff-7cfad0a0b04e44c6f6e568cfd6bce073202306ce4c0fb0c285ee1b7fce55836dR747. Not removing the __init__file now causes rather strange problems with imports.

Reply:
Fixed. Thanks.